readthedocs/builds/querysets.py

```diff
@@ -41,9 +42,27 @@ def _add_from_user_projects(self, queryset, user, admin=False, member=False):
             queryset = user_queryset | queryset
         return queryset
 
-    def public(self, user=None, project=None, only_active=True,
-               include_hidden=True, only_built=False):
-        queryset = self.filter(privacy_level=constants.PUBLIC)
+    def public(
+        self,
+        user=None,
+        project=None,
+        only_active=True,
+        include_hidden=True,
+        only_built=False,
+    ):
+        """
+        Get all allowed versions.
+
+        .. note::
+
+           External versions use the `Project.external_builds_privacy_level`
+           field instead of its `privacy_level` field.
+        """
+        queryset = self.filter(privacy_level=constants.PUBLIC).exclude(type=EXTERNAL)
+        queryset |= self.filter(
+            type=EXTERNAL,
+            project__external_builds_privacy_level=constants.PUBLIC,
+        )
```

Comment:
I'm a little worried about performance here -- we're adding a join and a few more queries on all our super common querysets. Can we not just change this on the `external` manager instead?

Reply:
Hmm, maybe, I'll need to take a look at where we are using those managers https://github.com/readthedocs/readthedocs.org/blob/ea574df48f7fd5d72e231df9f9cb1e9fa10dabf5/readthedocs/builds/models.py#L180-L184

Reply:
I think we can make this happen only when executing objects and external, as the internal manager already filters them out.

Reply:
I have used a partial class to pass a parameter indicating if the manager being used is for internal only versions.

Reply:
@stsewd It seems like it might be cleaned to put this on a `public` method in the `ExternalVersionManager` instead of on the base VersionQueryset, since it only applies there (eg. we only want to run it on `Version.external.public`, I believe). 

We should also do the same logic for the other querier (eg. on the Build)

Reply:
We also want to include this on the default manager, since we do have calls to `Version.objects.public`/`Build.objects.public`.
And I'm not sure about moving these to the manager, these are more filters, and all of our querysets are there, we would nee to move all of those. I was about to use inheritance instead of the partial, but then we need to create three more classes (parent, InternalQuerysetBase, InternalQueryset(SeetingsOverride)) and one more override in .com :/

Reply:
We don't need to worry about filtering external builds on `Version.objects.public` though right? They will already be filtered out, regardless of privacy level?

Reply:
I'm not sure if I understood what you mean :sweat_smile:.

- Version.objects.public -> external and internal versions. Internal versions are checked against `version.privacy_level`, and external versions are checked against `project.external_builds_privacy_level`.
- Version.internal.public -> internal versions only, checked against `version.privacy_level`
- Version.external.public -> external versions only, checked against `project.external_builds_privacy_level` (for this case we could omit one filter from the current implementation, I'll do that)

But anyway, Version.objects.public need to check for both external and internal as that manager is expected to include both types of versions.

Reply:
Ah yea, I was thinking `Version.internal.public`.